### PR TITLE
451001 - Anastasiya Lebedeva

### PR DESCRIPTION
**Pull request by lebedeanastasija**

### CODE REVIEW

File: TestTraceProgram/1htnfed5.gyl [hunk 1]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
```
Comment (RamanBut-Husaim) [inline TestTraceProgram/1htnfed5.gyl]: @lebedeanastasija Seems that the file should be removed from the source control.

File: TestTraceProgram/chkq03ex.ypt [hunk 2]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
```
Comment (RamanBut-Husaim) [inline TestTraceProgram/chkq03ex.ypt]: @lebedeanastasija Remove the file from the source control.

File: TestTraceProgram/kv5v1rqh.x1h [hunk 3]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
```
Comment (RamanBut-Husaim) [inline TestTraceProgram/kv5v1rqh.x1h]: @lebedeanastasija Remove the file from the source control.

File: TestTraceProgram/w4iifazu.3t3 [hunk 4]
```diff
@@ -0,0 +1,85 @@
+﻿using System;
```
Comment (RamanBut-Husaim) [inline TestTraceProgram/w4iifazu.3t3]: @lebedeanastasija Remove the file from the source control.

File: TracerAPI/033tnufe.ldj [hunk 5]
```diff
@@ -0,0 +1,90 @@
+﻿using System;
```
Comment (RamanBut-Husaim) [inline TracerAPI/033tnufe.ldj]: @lebedeanastasija Remove the file from the source control.

File: TracerAPI/ConsoleFormatter.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TracerAPI
+{
+    public class ConsoleFormatter: ITraceResultFormatter
+    {
+        private List<string> stack = new List<string>();
```
Comment (RamanBut-Husaim) [inline TracerAPI/ConsoleFormatter.cs]: @lebedeanastasija Why have you named this list `stack`?

File: TracerAPI/ConsoleFormatter.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                stack.Add(")");
+                foreach(Node child in node.Children){
+                    AddMethodsToTread(child, level);
+                }
+                Console.WriteLine(GetIndention(level) + stack.Last());
+                stack.RemoveAt(stack.Count - 1);
+            }
+        }
+
+        private string GetIndention(int level)
```
Comment (RamanBut-Husaim) [inline TracerAPI/ConsoleFormatter.cs]: @lebedeanastasija Take a look at the `string` constructor. It seems that it is possible to replace your method with a single call.

File: TracerAPI/ConsoleFormatter.cs (hunk 6 above)
Comment (RamanBut-Husaim) [inline TracerAPI/ConsoleFormatter.cs]: @lebedeanastasija As far as I see this field could be made `readonly`.

File: TracerAPI/Tracer.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        private static volatile Tracer instance = null;
+        private static readonly Object syncObj = new Object();
+       
+
+        private Tracer()
+        {
+            TraceInfo = new TraceResult();
+        }
+
+        public static Tracer Instance
```
Comment (RamanBut-Husaim) [inline TracerAPI/Tracer.cs]: @lebedeanastasija Take a look at other possible [singleton implementations](http://csharpindepth.com/Articles/General/Singleton.aspx)

File: TracerAPI/Tracer.cs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            Thread thread = Thread.CurrentThread;
+            int threadId = thread.ManagedThreadId; 
+
+            char  []separator = {' '};
+            string[] methodStrings = methodFrame.GetMethod().ToString().Split(separator);
+            int bracketIndex = methodStrings[1].IndexOf('(');
+            string methodName = methodStrings[1].Substring(0, bracketIndex);
+
+            int numberOfParams;
+            if (methodStrings.Count() == 2 && (methodStrings[1].IndexOf(')') -
```
Comment (RamanBut-Husaim) [inline TracerAPI/Tracer.cs]: @lebedeanastasija If you have `if` with `else` add curly braces `{}` to improve readability.

Comment (RamanBut-Husaim) [inline TracerAPI/Tracer.cs]: @lebedeanastasija Black magic...

File: TracerAPI/Tracer.cs [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+            string methodName = methodStrings[1].Substring(0, bracketIndex);
+            Tracer.AddStopInfo(threadId, methodName, stopTime);
+        }
+
+        public TraceResult GetTraceResult() 
+        {
+            return TraceInfo; 
+        }
+
+        public static void AddStartInfo(int threadId, string parentName,
```
Comment (RamanBut-Husaim) [inline TracerAPI/Tracer.cs]: @lebedeanastasija Why have you made this method `public`?

File: TracerAPI/Tracer.cs [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+                                        string methodName, int numberOfParams, string methodClassName, 
+                                        DateTime startTime)
+        {
+            Tree tree;           
+            tree = Tracer.Instance.GetTraceResult().GetTreeByThreadId(threadId);           
+            tree.AddNode(parentName, methodName, numberOfParams, methodClassName, startTime);
+        }
+
+        public static void AddStopInfo(int threadId, string methodName, DateTime stopTime)
```
Comment (RamanBut-Husaim) [inline TracerAPI/Tracer.cs]: @lebedeanastasija And this as well.

File: TracerAPI/Tree.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TracerAPI
+{
+    public class Tree
+    {
+        private Node root;
+        public Node Root { get { return root; } set { root = value; } }
```
Comment (RamanBut-Husaim) [inline TracerAPI/Tree.cs]: @lebedeanastasija Format into separate lines.

File: TracerAPI/dcmjx4vo.rfq [hunk 13]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
Comment (RamanBut-Husaim) [inline TracerAPI/dcmjx4vo.rfq]: @lebedeanastasija Remove the file from the source control.

File: TracerAPI/Node.cs [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    public class Node
+    {
+        public string MethodName { get; set; }
+        public string MethodClassName { get; set; }
+        public DateTime StartTime { get; set; }
+        public DateTime StopTime { get; set; }
+        public long WholeTime { get; set; }
+        public int NumberOfParameters { get; set; }
+
+        private List<Node> children = new List<Node>();
```
Comment (RamanBut-Husaim) [inline TracerAPI/Node.cs]: @lebedeanastasija `readonly`?

File: TracerAPI/Tracer.cs (hunk 9 above)
Comment (lebedeanastasija) [inline TracerAPI/Tracer.cs]: Okay, It will be a white magic next time

File: TracerAPI/ConsoleFormatter.cs (hunk 6 above)
Comment (lebedeanastasija) [inline TracerAPI/ConsoleFormatter.cs]: Because  I work with it like with stack.

Comment (lebedeanastasija) [inline TracerAPI/ConsoleFormatter.cs]: Because I  didn't know about C# Stack<T> class  and  worked with list like with stack.

Comment (lebedeanastasija) [inline TracerAPI/ConsoleFormatter.cs]: Why should I make this field readonly? I change it in methods.

Comment (RamanBut-Husaim) [inline TracerAPI/ConsoleFormatter.cs]: @lebedeanastasija OK.

Comment (RamanBut-Husaim) [inline TracerAPI/ConsoleFormatter.cs]: @lebedeanastasija You are changing only its internal state. Be ready to answer what is `readonly` and describe its semantics.

File: TracerAPI/Node.cs (hunk 14 above)
Comment (lebedeanastasija) [inline TracerAPI/Node.cs]: But I changes it in the program.

File: TracerAPI/Tracer.cs (hunk 10 above)
Comment (lebedeanastasija) [inline TracerAPI/Tracer.cs]: I really don't now.. May be the reason is "why not

Comment (lebedeanastasija) [inline TracerAPI/Tracer.cs]: ".

Comment (lebedeanastasija) [inline TracerAPI/Tracer.cs]: It should me private of course.

File: TracerAPI/Tracer.cs [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+            Tracer.AddStopInfo(threadId, methodName, stopTime);
+        }
+
+        public TraceResult GetTraceResult() 
+        {
+            return TraceInfo; 
+        }
+
+        private static void AddStartInfo(int threadId, string parentName,
```
Comment (RamanBut-Husaim) [inline TracerAPI/Tracer.cs]: @lebedeanastasija Why have you made these methods `static` with accessing trace result through tracer singleton?

File: TracerAPI/Tracer.cs [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+
+        private static void AddStopInfo(int threadId, string methodName, DateTime stopTime)
+        {
+            Tree tree;
+            tree = Tracer.Instance.GetTraceResult().GetTreeByThreadId(threadId);
+            tree.CompleteNode(methodName, stopTime);
+        }
+
+        private bool HasOneParameter(string parametersString)
+        {
```
Comment (RamanBut-Husaim) [inline TracerAPI/Tracer.cs]: @lebedeanastasija The internal content of this method could be replaced with one line
`return parametersString.IndexOf(')') - parametersString.IndexOf('(') > 1;`

File: TracerAPI/Tracer.cs [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+            if (parametersParts.Count() == 1)
+            {
+                if(HasOneParameter(parameters))
+                {
+                    numberOfParams = 1;
+                }
+            }
+            else
+            {
+                if(parametersParts.Count() > 1)
```
Comment (RamanBut-Husaim) [inline TracerAPI/Tracer.cs]: @lebedeanastasija Why have you used curly brackets above and omitted them right here?

File: TracerAPI/Tree.cs [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+                             node.StartTime.Minute * 1000 * 60);
+        }
+        
+
+        public Node GetParentNodeByName(string parentMethodName, Node tempNode)
+        {
+            Node parentNode = null;
+            if(Root != null)
+            { 
+                if(parentMethodName == tempNode.MethodName)
```
Comment (RamanBut-Husaim) [inline TracerAPI/Tree.cs]: @lebedeanastasija The issue with curly brackets.

File: TracerAPI/XMLFormatter.cs [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+                Thread.Add(Method);
+
+                Method.SetAttributeValue("name",traceResult[key].Root.MethodName);
+                Method.SetAttributeValue("parameters", traceResult[key].Root.NumberOfParameters);
+                Method.SetAttributeValue("class-name", traceResult[key].Root.MethodClassName);
+                Method.SetAttributeValue("time", traceResult[key].Root.WholeTime);
+
+                AddChildrenToXElement(Method, traceResult[key].Root);
+            }
+            Root.Save("ThreadsTree.xml");
```
Comment (RamanBut-Husaim) [inline TracerAPI/XMLFormatter.cs]: @lebedeanastasija Move file name to an instance field and accept it via constructor.

**Pull request merged**